metrics/src/recorder.rs

```diff
@@ -1,13 +1,105 @@
+use self::cell::RecorderOnceCell;
 use crate::{Counter, Gauge, Histogram, Key, KeyName, Unit};
 use core::fmt;
-use core::sync::atomic::{AtomicUsize, Ordering};
 
-static mut RECORDER: &'static dyn Recorder = &NoopRecorder;
-static STATE: AtomicUsize = AtomicUsize::new(0);
+mod cell {
+    use super::{Recorder, SetRecorderError};
+    use std::cell::UnsafeCell;
+    use std::sync::atomic::{AtomicUsize, Ordering};
 
-const UNINITIALIZED: usize = 0;
-const INITIALIZING: usize = 1;
-const INITIALIZED: usize = 2;
+    // FIXME: This can't be a const new function because trait objects aren't allowed in const fns
+    // This was stabilized in 1.61, so it can be cleaned up when it becomes the MSRV
+    #[allow(clippy::declare_interior_mutable_const)]
+    pub const INIT: RecorderOnceCell = RecorderOnceCell {
+        recorder: UnsafeCell::new(None),
+        state: AtomicUsize::new(UNINITIALIZED),
+    };
+
+    /// The global Recorder instance with a `once_cell`-like API
+    pub struct RecorderOnceCell {
+        recorder: UnsafeCell<Option<&'static dyn Recorder>>,
+        state: AtomicUsize,
+    }
+
+    /// The recorder is uninit and can be set
+    const UNINITIALIZED: usize = 0;
+    /// The recorder is currently being initialized
+    const INITIALIZING: usize = 1;
+    /// The recorder has been initialized successfully and can be read
+    const INITIALIZED: usize = 2;
+
+    impl RecorderOnceCell {
+        #[cfg(atomic_cas)]
+        pub fn set(&self, recorder: &'static dyn Recorder) -> Result<(), SetRecorderError> {
+            // Acquire the lock because the write below must not be reordered above the CAS
+            match self.state.compare_exchange(
+                UNINITIALIZED,
+                INITIALIZING,
+                Ordering::Acquire,
+                Ordering::Relaxed,
+            ) {
+                Ok(UNINITIALIZED) => {
+                    unsafe {
+                        // SAFETY: Access is unique because we CASed the state to INITIALIZING above
+                        self.recorder.get().write(Some(recorder));
+                    }
+                    // Release the lock, others can now read it - but not write
+                    self.state.store(INITIALIZED, Ordering::Release);
+                    Ok(())
+                }
+                _ => Err(SetRecorderError(())),
+            }
+        }
+
+        /// # Safety
+        /// The caller must guarantee that no reader has read the state before we do this and then
+        /// reads the recorder after another writer has written to it after us.
```

Comment:
This needs an actual high-level description, not just the safety section. I realize it's not actually part of the crate's public API, but if it's going to have documentation at all, it needs to be well-formed.

Something as simple as `/// Clears the currently installed recorder.` would suffice.